controllers/cluster_controller.go

```diff
@@ -123,6 +123,15 @@ func (r *ClusterReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr e
 		return ctrl.Result{}, err
 	}
 
+	// Add finalizer first if not exist to avoid the race condition between init and delete
```

Comment:
@SataQiu 
Dummy question for my understanting: is there is a specific reason for patching finalizers in the main reconcile loop vs adding finalizers and returning for requeue in reconcileNormal (so we can rely on the same patch method)

Reply:
@fabriziopandini  It's a good pattern to avoid the race condition between init and delete. Because we always check the `Finalizer` and patch it before the `reconcileDelete` logic. Maybe I'm missing something. Need more help.

Reply:
@SataQiu I agree it is a good pattern.
My comment was about an implementation detail, because:
1. I think we should keep the main reconcile loop as simple as possible
2. In the main reconcile loop we are already calling defer patch, so, if possible, I would like to avoid implementing a separate call to patch / patchFinalizer.
3. We are already checking/adding finalizers at the beginning of reconcileNormal (we are just not returning now)

However, said that I would aim to simplicity, so far I don't have strong preferences, and so I'm trying to understand if there is a specific reason for selecting the current approach.

Reply:
Did you mean moving it down so we can reuse the same patch finalizer that's being initialized on line 121? That makes sense to me. 

That said, we should still keep the finalizer part in the `Reconcile` method vs adding it in `reconcileNormal` because there we've had a case before where the object was deleted too quickly after creation and the finalizer was never written/patched. To ensure a proper cleanup, we'll need to add a finalizer first and then call `reconcileNormal` or `reconcileDelete`.

Reply:
I was thinking to have something similar in reconcileNormal
```
if add := controllerutil.AddFinalizer(mp, expv1.MachinePoolFinalizer); add {
    return nil
}
```
So basically patching finalized happens before starting any operation, but at the same time, it re-uses the same defer Patch

Reply:
The util part seems fine, not sure how necessary — but it needs to be in both reconcile normal and delete part of the codebase, or in the main Reconcile method. Current changes look good to me

Reply:
@vincepri can you clarify what you mean by it needing to be in both the reconcile normal and delete parts of the codebase?

I wouldn't think we would need to Add the finalizer if we are in the delete part of the workflow, and would only want to remove the finalizer after we finish taking action on the resource to prevent the api server from cleaning up the resource prior to finishing.

I also don't think we need to worry about a race condition between adding the finalizer (if doing it via short-circuit prior to any other actions in `reconcileNormal` and skipping any cleanup tasks in `reconcileDelete` if the finalizer is not present.

I do think that doing it in `Reconcile` would help to prevent potential programming errors if any logic sneaks in prior to adding the finalizer and returning early in `reconcileNormal`, but not necessarily that is a good enough reason to keep it out of `reconcileNormal`

Reply:
We've had an issue before where for example let's say you quickly create/delete a Machine, the linked object wouldn't be deleted (i.e. bootstrap and instrastructure refs) leaving them around. Making sure that the finalizer is always added if the object is being reconciled gives the possibility to the deletion loop to do proper cleanups and remove the finalizer when done.

Reply:
given above context, I'm ok with the current approach.